### ISSUE TITLE
Separate coala-aspects project

### ISSUE THREAD

**Issue by userzimmermann** [coala/coala#4038]
aspects are just so fundamental that they should also get their separate repository and python package/distribution ;) And now seriously:

I think we could really speed up the aspects development in a separate project. And when the number of aspects grow and people get interested in contributing new aspect definitions, it gets messy to keep them nested in `coalib.bearlib.aspects` and together with so many other non-aspects related issues and pull requests in the main **coala** project. I should feel more like bear development in the separate **coala-bears** project.

So I propose creating a **coala/coala-aspects** project with its own distribution and a `coala_aspects` top-level package, in which we move the current content of `coalib/bearlib/aspects`, and making the **coala** distribution depend on it.

@sils @jayvdb @coala/aspects-developers

**Comment by Techievena** [comment]
Yeah @userzimmermann I agree to this. Not exactly like the __coala-bears__ project because that is independent of __coala__ but there should be a separate __aspects__ repository like the __package-manager__ or __coala-utils__ repository. As aspect-classes are extendable therefore there must be a separate repo for __aspects__.

**Comment by Techievena** [comment]
@userzimmermann But I am quite unsure with the python distribution thing. We can distribute bears grouped by aspects but why distribute aspect-classes separately? I mean if I have a bear installed without having its aspect-class it can be a problem. I think it is better to have all of them at once.

**Comment by adhikasp** [comment]
@Techievena 
> making the coala distribution depend on it
I'm pretty sure it mean coala pip package is dependant of coala-aspect. So we don't need to explicitly install coala-aspect itself.

**Comment by pratyushprakash** [comment]
I support this idea. I also think we should organize the aspects a little more instead of directly moving them to the new repository.

**Comment by Techievena** [comment]
@adhikasp 
>So we don't need to explicitly install coala-aspect itself
That's why there is no need for package distribution of aspect-classes.

**Comment by userzimmermann** [comment]
@Techievena The comparison with **coala-bears** was not related to dependency. As @adhikasp correcly pointed out again, **coala** should of course depend on **coala-aspects** and not the other way around, looking basically like:

```
coala-bears
|         |
V         |
coala     |
|   |     |
|   V     V
|   coala-aspects
|   |
V   V
coala-utils
```

But without creating a **coala-aspects** distribution, there won't be a way to distribute it ;)

**CC** @sils

**Comment by jayvdb** [comment]
I am having difficulty seeing the benefit of splitting the aspects code out of coala , to warrant the costs associated with doing it.

The module name shortening argument is not nearly good enough.  There are other ways to shorten the module names.

Or is this mostly about the aspects team having its own release schedule that isn't tied to coala?

Who (other than coala) would use `coala-aspects`?

One way to 'reduce' the costs of having a separate coala-aspects is to require that it does not (ever!) add any dependencies without prior explicit approval from the release team, so they have time to check the dependency isn't going to break coala.

**Comment by userzimmermann** [comment]
@jayvdb 
> Or is this mostly about the aspects team having its own release schedule that isn't tied to coala?

Another good point! Thanks :)

I also found an additional advantage: We can then completely move to `pytest` style for our tests. And don't have to fight with people anymore, who want to convert our tests to `unittest` style ;D

> Who (other than coala) would use coala-aspects?

**coala-bears** ;) It's just like with **coala-utils**

It just gives the aspects team more flexibility. We are not so many people and need to quickly get forward with aspects, and need a shorter release cycle to be able to quickly migrate bears and implement new bears with newly defined aspects. We stand anyway a bit out from other coala developers/teams, since we are focused on a very specific experimental part of coala, which is not yet needed for coala's current functionality, which will be integrated into coala and bears in parallel to the current internal mechanisms while maintaining full backwards compatibility, and which is rarely touched by other contributors so far...

@coala/aspects-developers Find more good reasons please!

**Comment by jayvdb** [comment]
coala-utils is full of routines which are usable by other projects.  If I go back to working on pywikibot, I would try to integrate it there.

Who other than coala* (any coala project) could use it?

**Comment by sils** [comment]
Actually I *could* see linters using the coala aspects without fully wanting to rely on coala itself. If we decouple it from coala completely we might get more linters to adapt our technologies and maybe set a standard or so. Then again that'll only work for python linters with aspects.

@userzimmermann I don't think untying aspects release schedule from coala's makes sense. coala and coala-bears are kept in sync for the same reason.

I think the main point here would be giving the aspects team an independence so they can iterate quickly right now and are not so much affected by coala's deprecation etc. policies, make a clear cut. Aspects *are* experimental and getting them to work sometime soon would be nice; moving fast and breaking things is ok for aspects if they fix it on release time. Essentially: "the bigger the team the bigger your problem" so splitting up can result in a significant raise of productivity I think.

That being said I'm not sure what's the best course of action.

**Comment by Techievena** [comment]
@sils The release cycle should be made more frequent because it won't be possible to write bears requiring new aspect-classes until we have those aspect-classes definition.

**Comment by sils** [comment]
@Techievena in contrary this'll be simpler if aspects remain in coala since coala auto releases development versions and we update the brlears to the latest Dev from coala amaster when needed

**Comment by Techievena** [comment]
Can't we have the same thing implemented in the aspects repository. Because this single thing doesn't outweighs the other benefits of having a new repository for aspects.

**Comment by Techievena** [comment]
@userzimmermann reply to https://github.com/coala/coala/issues/4038#issuecomment-290997548 
@sils Aren't there any other mechanisms to directly use aspects from the git repository itself?
I am still having doubts about packaging and having releases for aspects repository. I don't think a bunch of aspect-classes will be of any help to the outside sources.

**Comment by adhikasp** [comment]
I am against this idea, at least in this near future.

In my view, aspect project needs to **integrate aspect into coala core** and then **expand definition and maybe other aspects features**. 

For the integration phase, I think the separation would make more hassle than benefit. The effort spent to create the distribution package, creating and maintaining dependency between coala and coala-aspects, etc is better spent to integrate the aspect itself. Also, I am picturing some scenario when an aspect PR to coala was blocked because it's need some modification in coala-aspect, thus unnecessarily stalling time. If we don't separate, at least we can pass that as 1 PR and make the process faster.

Also before integration complete, aspect could not run anyway. So the flexible release schedule is not really beneficial.

After aspect could actually functional, we could get a better picture whether to really separate coala and coala-aspect or not.

**TLDR** separate coala-aspect (if needed) after integration into coala core, but not before it.

**Comment by Makman2** [comment]
Though I lost track of what is the final goal, aspects seem to be quite bound to coala, so moving it into a separate repo might just make dependency problems.

**Comment by userzimmermann** [comment]
@sils Regarding https://github.com/coala/coala/issues/4038#issuecomment-293945348 : But bears using new aspects couldn't go into production before the next **coala** release containing those new aspects. Also, it's not so straightforward - especially for not so experienced developers - to migrate or create new bears with new aspects based on development releases of **coala**

**Comment by Makman2** [comment]
@userzimmermann actually that's the current workflow when we have new features. If there's a new feature and coala-bears does use it, the minimum requirement is the dev-version introducing the new feature. Actually I don't think it's so inconvenient^^

>  But bears using new aspects couldn't go into production before the next coala release containing those new aspects

The same does hold for an aspects-release. If it's brand new, we don't release it immediately normally. People have to use the dev-version or some kind of pre-release/alpha/beta (if we wanna do that).

**Comment by userzimmermann** [comment]
@Makman2 

> The same does hold for an aspects-release. If it's brand new, we don't release it immediately normally. People have to use the dev-version or some kind of pre-release/alpha/beta (if we wanna do that).

But we can release it earlier and more often than **coala** :)

**Comment by userzimmermann** [comment]
@Makman2 

> Though I lost track of what is the final goal, aspects seem to be quite bound to coala, so moving it into a separate repo might just make dependency problems.

It's just one more **coala-utils**-like dep to **coala** and **coala-bears**

**Comment by userzimmermann** [comment]
@Techievena 

> Aren't there any other mechanisms to directly use aspects from the git repository itself?
I am still having doubts about packaging and having releases for aspects repository. I don't think a bunch of aspect-classes will be of any help to the other projects outside coala. That will also solve the above issue of having new aspect-classes for new bears.

That's actually a very interesting approach!

@sils Seriously! What about some bears/aspects online repository with some custom installation management instead of bundling everything into single python packages? That could also speed up bear development by external contributors a lot...

**Comment by Makman2** [comment]
It's possible to specify commits inside pip requirements list if I recall correctly. Somehow you have to use the git extension, looks something like `git+URL-commit` or so...

**Comment by jayvdb** [comment]
> Seriously! What about some bears/aspects online registry with some custom installation management instead of bundling everything into single python packages?

@userzimmermann , that is already happeneing. We are using pypi as an online registry.  See https://github.com/coala/coala-bears/blob/master/bears/generate_package.py and [pypi search 'coala bear'](https://pypi.python.org/pypi?%3Aaction=search&term=coala+bear&submit=search) and [cib](https://gitlab.com/coala/cib)

That is not an argument for separate repos for each bear.  We could break the bears repos into subsets, but splitting them into one repo per bear means it is more work to create a new bear, as it requires a new repo, with new CI setup, and how to get review for a WIP bear, given that some bears stay in WIP mode for months because of the initial authors level of commitment.

Back to a separate repo for aspects, creating a separate repo for coalangs, docstyle definitions, and aspects is a reasonable idea in the future. (likely a single repo for all of them combined, as they are all inter-related metadata).  That should happen once they are well defined and approaching stability!

**Comment by Makman2** [comment]
> Back to a separate repo for aspects, creating a separate repo for coalangs, docstyle definitions, and aspects is a reasonable idea in the future. (likely a single repo for all of them combined, as they are all inter-related metadata). That should happen once they are well defined and approaching stability!

totally agreeing :+1:

So I also vote for doing that later when we see that this is reasonable. Currently aspects aren't really implemented yet, keeping things in the same repo eases using this feature for now.
